pyccel/ast/core.py

```diff
@@ -2468,126 +2483,19 @@ def _eval_is_positive(self):
         #we do this inorder to infere the type of Pow expression correctly
         return self.is_real
 
-class DottedVariable(AtomicExpr, sp_Boolean, PyccelAstNode):
+class DottedVariable(Variable):
 
     """
     Represents a dotted variable.
     """
 
-    def __new__(cls, lhs, rhs):
-
-        if PyccelAstNode.stage != 'syntactic':
-            if not isinstance(lhs, (
-                Literal,
-                Variable,
-                Symbol,
-                IndexedVariable,
-                IndexedElement,
-                IndexedBase,
-                Indexed,
-                Function,
-                DottedVariable,
-                )):
-                raise TypeError('Expecting a Variable or a function call, got instead {0} of type {1}'.format(str(lhs),
-                                str(type(lhs))))
-
-            if not isinstance(rhs, (
-                Variable,
-                Symbol,
-                IndexedVariable,
-                IndexedElement,
-                IndexedBase,
-                Indexed,
-                FunctionCall,
-                Function,
-                )):
-                raise TypeError('Expecting a Variable or a function call, got instead {0} of type {1}'.format(str(rhs),
-                                str(type(rhs))))
-
-        return Basic.__new__(cls, lhs, rhs)
-
-    def __init__(self, lhs, rhs):
-        if self.stage == 'syntactic':
-            return
-        self._dtype     = rhs.dtype
-        self._rank      = rhs.rank
-        self._precision = rhs.precision
-        self._shape     = rhs.shape
-        self._order     = rhs.order
+    def __init__(self, *args, lhs, **kwargs):
+        Variable.__init__(self, *args, **kwargs)
```

Comment:
I'm not sure about this, we should be able to do something like `(a-b).sum()`

Reply:
I guess it's the same problem for `DottedFunctionCall`, when we have something like `(2*b).sum()`

Reply:
the first element can be an expression , the rest should be variables or variables + function call

Reply:
I don't see the problem. I don't impose any restrictions on the lhs. So for `(a-b).shape` the dotted variable would contain the shape part with
shape=1
rank=0
dtype=int
And the lhs (which is visited separately) would contain `PyccelMinus(a,b)`

Reply:
how about the name ?

Reply:
We can't test the examples you mention for 2 reasons
1. Pyccel does not handle magic methods
2. `shape` and `sum` are special cases which would not be translated to Dotted objects anyway

But perhaps the following test (I have not tried it yet) would be enough to convince you:
```
class A:
    def __init__(self):
        self.x =3

    def f(self):
        return False

def get_A():
    return A()

a = get_A().x
b = get_A().f()
```
?

Reply:
> how about the name ?

The name would be `shape` (although of course shape would actually be converted to `PyccelAstNode`). The name is the name of the class variable (this seems to be necessary to easily collect the object from the `ClassDef`). The printer then just does:
```
self._print(expr.lhs) + ' % ' + self._print(expr.name)
```
However you are right that brackets would be needed if the lhs is not a singleton

Reply:
yes, I agree we don't have those cases yet, but we will handle magic methods at a certain point or the example that you gave, we just need to think about them for now and decide what to do for later , so that we won't change the dotted variable once more

Reply:
you probably need to do something like
```python
def __init__(self, *args):
    self.clone(args[-1])
```

Reply:
but then there is this case `a.b().c.d().e` , does it make sense to seperate the dotted function call from the dotted variable , I mean it should still work fine with the seperation, but do we need it ?

Reply:
The example above is now working. I had to make a few changes to the parser so it would parse correctly, but no changes to `DottedVariable` or `DottedFunctionCall`.
I have added it to the tests

Reply:
ah perfect

Reply:
> but then there is this case `a.b().c.d().e` , does it make sense to seperate the dotted function call from the dotted variable , I mean it should still work fine with the seperation, but do we need it ?

I think it helps as much as having a separation between Variable and FunctionCall. The problem with keeping everything in the same object is that every time you might encounter a DottedVariable (which is everytime you might encounter a Variable) you would have to check the type and collect the Variable.
Consider
```
a.f()+a.x
```
with the inheritance here, this is handled in exactly the same way as
```
f()+x
```
which means that we don't have to maintain classes separately. But if they are not separated then you need to check all the arguments to `PyccelAdd` to check for Dotted types and then you need to retrieve the object with the relevant information. This would need doing for every class in pyccel which needs to calculate its dtype (and there are a lot)

Reply:
ah okay I see